tools/autograd/derivatives.yaml

```diff
@@ -129,73 +119,197 @@
   self: grad * tensor
   tensor: grad * self
 
-- name: eig
-- name: eq
-- name: equal
-- name: exp
+- name: eig(Tensor self, bool eigenvectors)
+  self: not_implemented("eig")
+
+- name: eq(Tensor self, Scalar value)
+  self: zeros_like(self)
+
+- name: eq(Tensor self, Tensor other)
+  self: zeros_like(self)
+  other: zeros_like(other)
+
+- name: equal  # fallthrough
+
+- name: erf(Tensor self)
+  self: 2.0 / sqrt(M_PI) * exp(-(self.pow(2))) * grad
+
+- name: erfinv(Tensor self)
+  self: 0.5 * sqrt(M_PI) * exp(self.erfinv().pow(2)) * grad
+
+- name: exp(Tensor self)
+  self: grad * result
 
 - name: expand(Tensor self, IntList size)
   self: reduce_to(grad, self.sizes())
+  __view__: True
+
+- name: eye  # fallthrough
 
-- name: eye
-- name: fill
-- name: floor
-- name: fmod
-- name: frac
-- name: gather
-- name: ge
-- name: gels
-- name: geometric
-- name: geqrf
+- name: fill(Tensor self, Scalar value)  # FIXME
+
+- name: floor(Tensor self)
+  self: zeros_like(grad)
+
+- name: fmod(Tensor self, Scalar value)
+  self: grad
+
+- name: fmod(Tensor self, Tensor other)
+  self: grad
+  other: 'not_implemented("fmod: other")'
+
+- name: frac(Tensor self)
+  self: grad
+
+- name: gather(Tensor self, int64_t dim, Tensor index)
+  self: grad.type().zeros(self.sizes()).scatter_add_(dim, index, grad)
+
+- name: ge(Tensor self, Scalar value)
+  self: zeros_like(self)
+
+- name: ge(Tensor self, Tensor other)
+  self: zeros_like(self)
+  other: zeros_like(other)
+
+- name: gels(Tensor self, Tensor A)
+  self: not_implemented("gels")
+  A: not_implemented("gels")
+
+- name: geometric(Tensor self, double p, Generator Generator)
+  self: zeros_like(grad)
+
+- name: geqrf(Tensor self)
+  self: not_implemented("geqrf")
 
 - name: ger(Tensor self, Tensor vec2)
   self: grad.mv(vec2)
   vec2: grad.t().mv(self)
 
-- name: gesv
+- name: gesv(Tensor self, Tensor A)
+  self: std::get<0>(gesv(grad, A.t()))
+  A: -at::mm(std::get<0>(gesv(grad, A.t())), solution.t())
```

Comment:
nit: if we want to reuse intermediate computations (like the `gesv(grad, A.t())`), the recommended way will be to write the derivative in C++ and not rely on the yaml, right?

Reply:
Yes and it's not currently supported for functions in derivatives.yaml. I'd like to write:

`derivatives.yaml`:
```yaml
name: gesv(Tensor self, Tensor A)
self, A: gesv_backward(grad, A, solution)
```

`tools/autograd/templates/Functions.cpp`:
```c++
std::tuple<Tensor, Tensor> gesv_backward(...) {
 ...
}
```

This will require some more modifications to `gen_variable_type.py`.

Reply:
As you notices, `gen_variable_type` starts getting complicated, which is a good sign that we should stop extending the language of `derivatives.yaml`. I doubt that `gesv` is a significant bottleneck for models, so let's just keep it implemented in Python. That file should only be used as an easy method to port simple functions to C++

Reply:
We're going to need it for other things too like prelu double backwards

Reply:
We'll have in `derivatives.yaml`:

```yaml
- name: prelu_backward(Tensor grad_output, Tensor input, Tensor weight, std::array<bool, 2> output_mask):
  grad_output, input, weight: prelu_double_backward(grads, grad_output, input, weight)
```

In some C++ file (possibly Functions.cpp):

```c++
std::tuple<Tensor, Tensor, Tensor> prelu_double_backward(TensorList grads, const Tensor & grad_output, const Tensor & input, const Tensor & weight) {
 // implemenation from PReLUBackward.backward
}
```

(It's possible that the double backwards for `grad_output` could be "unfused" from the rest of the computation)

There's two pieces of functionality we need to support this:
1. Support multiple values on the lhs in derivatives.yaml (this should be a small change)
2. Support taking multiple "grad_outputs" in backwards functions. Currently, all our autogenerated backwards functions assume as single grad_output as input.